Add crypto support to TI cc23x0 SoC

jpanisbl opened the pull request:
This series adds crypto support to TI cc23x0 SoC.

Datasheet: https://www.ti.com/lit/ds/symlink/cc2340r5.pdf

jpanisbl committed:
drivers: crypto: Add support for cc23x0 AES

Add support for AES module to cc23x0 SoC. The driver supports the
following modes:
- ECB encryption only (since decryption is not supported by the HW)
- CTR
- CCM

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>
Signed-off-by: Julien Panis <jpanis@baylibre.com>

jpanisbl committed:
dts: arm: ti: cc23x0: Add crypto support

Add support for AES module to cc23x0 SoC.

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>
Signed-off-by: Julien Panis <jpanis@baylibre.com>

jpanisbl committed:
boards: ti: lp_em_cc2340r5: Add crypto support

Enable AES module.

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>
Signed-off-by: Julien Panis <jpanis@baylibre.com>

valeriosetti commented on drivers/crypto/crypto_cc23x0.c:
```diff
@@ -0,0 +1,720 @@
+
+	AESSetIMASK(CRYPTO_CC23_INT_MASK);
+
+	data->in_use = false;
```
Wouldn't it be more thread-safe to use a mutex here and then `k_mutex_lock/k_mutex_unlock` in `crypto_cc23x0_session_setup/crypto_cc23x0_session_free` respectively?

jpanisbl replied:
Absolutely.

valeriosetti reviewed (changes requested):
The overall implementation looks good to me. I only left few suggestions for improvements (some of which are also optional)

jpanisbl commented on drivers/crypto/crypto_cc23x0.c:
```diff
@@ -0,0 +1,720 @@
+	.query_hw_caps = crypto_cc23x0_query_caps,
+};
+
+static struct crypto_cc23x0_data crypto_cc23x0_dev_data;
```
Thank you for the review Valerio.
I use `DT_INST_FOREACH_STATUS_OKAY()` for drivers that usually take care of several instances, but I thought it was not worth implementing here. Also, there are pending patches (not submitted upstream yet) to add DMA mode and PM support to this driver, and I'm not sure it's worth decreasing the readability in favor of a multi-instance approach that is not actually needed.
That said, I may be wrong. So I'll show you these pending patches. And if you still think it's worth using `DT_INST_FOREACH_STATUS_OKAY()`, I'll do.

valeriosetti replied:
Don't worry, there are already other drivers doing the same in this folder (actually most of the other drivers do exactly the same), so I think this is accepted and you can ignore my comment ;)

jpanisbl replied:
OK. All your other feedbacks are implemented.

jpanisbl commented on drivers/crypto/crypto_cc23x0.c:
```diff
@@ -0,0 +1,720 @@
+		out_bytes_processed += AES_BLOCK_SIZE;
+	}
+
+	/* Avoid triggering a spurious encryption upon reading the final output */
+	AESClearAUTOCFGTrigger();
+
+	/* Wait for AES operation completion */
+	ret = k_sem_take(&data->aes_done, CRYPTO_CC23_OP_TIMEOUT);
+	if (ret) {
+		goto cleanup;
+	}
+
+	LOG_DBG("AES operation completed");
+
+	/* Read final output */
+	AESReadTXT(&pkt->out_buf[out_bytes_processed]);
+	out_bytes_processed += AES_BLOCK_SIZE;
```
I need to double-check, but at first glance that looks possible indeed.

jpanisbl commented on drivers/crypto/crypto_cc23x0.c:
```diff
@@ -0,0 +1,720 @@
+
+	for (i = 0 ; i < sizeof(msg_len) ; i++) {
+		byte = CCM_CC23_BYTE_GET(i, msg_len);
+		memcpy(&b0[AES_BLOCK_SIZE - 1 - i], &byte, 1);
```
Your question is not dumb at all. My `memcpy` here is dumb. :-)
I probably started with more bytes, changed it, and ended up being a bit "code-blind".

jpanisbl committed:
samples: drivers: crypto: Add support for cc23x0

For the CCM mode, increase slightly the output buffer size because
the cc23x0 directly fills this buffer with 16-byte blocks.
This prevents from doing unnecessary memcpy that would have an
impact on the driver's performance.

Signed-off-by: Julien Panis <jpanis@baylibre.com>

valeriosetti reviewed (dismissed):
Thanks for addressing my comments. LGTM :)

ceolin commented on drivers/crypto/crypto_cc23x0.c:
```diff
@@ -0,0 +1,705 @@
+	}
+
+	/* Load key */
+	AESWriteKEY(ctx->key.bit_stream);
```
I am wondering if you shouldn't protect this call with the semaphore. What happens if another thread changes the key while there is an operation happening ?

jpanisbl replied:
Thank you for your time Flavio.
You're right, that's something that must be addressed. That's also the case for the whole encrypt/decrypt operation (not only for the key).

jpanisbl replied:
Also, I read more carefully the documentation about "[mutexes](https://docs.zephyrproject.org/latest/kernel/services/synchronization/mutexes.html) VS [semaphores](https://docs.zephyrproject.org/latest/kernel/services/synchronization/semaphores.html)". and it seems to me that I should replace the existing mutex (currently used to protect the session) with a semaphore. The mutex I'm using to protect the session may not do the trick, since it can be locked twice by the same thread. I will double-check with @valeriosetti .
IOW, there will probably be 2 semaphores: 1 for the session, and 1 for the access to HW.

valeriosetti replied:
After some internal discussion we agreed to:

- remove the session semaphore/mutex. Rationale: we don´t care if multiple thread start multiple sessions; we only care not to use AES HW at the same time.
- change `device_sem` to be a mutex and set it to protect any access to the AES HW (mutex over semaphore in this case because mutexes are thread-aware).
- keep the aes_done of course for HW synchronization.

This is the same as `crypto_si32.c` and `crypto_stm32.c`.

ceolin reviewed (dismissed):
I have only one concern about the semaphore usage across crypto operations because I don't know how the hw operates.

From the driver interface perspective it looks good.

jpanisbl commented:
As a complement, using the sample that is part of this PR:
```
*** Booting Zephyr OS build 3d850a41df11 ***
I: Cipher Sample
I: ECB Mode
D: AES operation completed
I: Output length (encryption): 16
**I: ECB mode ENCRYPT - Match**
**E: ECB decryption not supported by the hardware**
I: CBC Mode
E: Unsupported mode
I: CTR Mode
D: AES operation completed
D: AES operation completed
D: AES operation completed
D: AES operation completed
I: Output length (encryption): 64
**I: CTR mode ENCRYPT - Match**
D: AES operation completed
D: AES operation completed
D: AES operation completed
D: AES operation completed
I: Output length (decryption): 64
**I: CTR mode DECRYPT - Match**
I: CCM Mode
D: Compute CMAC
D: AES operation completed (block 0)
D: AES operation completed (block 1)
D: AES operation completed (data)
D: AES operation completed (data)
D: Encrypt data
D: AES operation completed
D: AES operation completed
D: Encrypt tag
D: AES operation completed
I: Output length (encryption): 23
**I: CCM mode ENCRYPT - Match**
D: Decrypt data
D: AES operation completed
D: AES operation completed
D: Compute CMAC
D: AES operation completed (block 0)
D: AES operation completed (block 1)
D: AES operation completed (data)
D: AES operation completed (data)
D: Encrypt tag
D: AES operation completed
D: Check tag
I: Output length (decryption): 23
**I: CCM mode DECRYPT - Match**
I: GCM Mode
E: Unsupported mode
```

ceolin reviewed (approved):
Thanks guys !
